pkg/resource/deploy/step.go

```diff
@@ -84,17 +85,16 @@ func NewSameStep(plan *Plan, reg RegisterResourceEvent, old *resource.State, new
 
 func (s *SameStep) Op() StepOp           { return OpSame }
 func (s *SameStep) Plan() *Plan          { return s.plan }
-func (s *SameStep) Type() tokens.Type    { return s.old.Type }
-func (s *SameStep) Provider() string     { return s.old.Provider }
-func (s *SameStep) URN() resource.URN    { return s.old.URN }
+func (s *SameStep) Type() tokens.Type    { return s.new.Type }
+func (s *SameStep) Provider() string     { return s.new.Provider }
+func (s *SameStep) URN() resource.URN    { return s.new.URN }
 func (s *SameStep) Old() *resource.State { return s.old }
 func (s *SameStep) New() *resource.State { return s.new }
 func (s *SameStep) Res() *resource.State { return s.new }
 func (s *SameStep) Logical() bool        { return true }
 
 func (s *SameStep) Apply(preview bool) (resource.Status, StepCompleteFunc, error) {
-	// Retain the URN, ID, and outputs:
-	s.new.URN = s.old.URN
+	// Retain the ID, and outputs:
```

Comment:
When does `s.new.URN` end up getting set, now?

Reply:
If you look above at the constructor, it was already asserting:

```
contract.Assert(new.URN != "")
```

It was necessary to establish a new URN early in the step generator to decide whether there is an old at all - which is the first thing step generator does before creating (and then applying) these steps.

https://github.com/pulumi/pulumi/blob/042f904547bc8155f37e956be9e40a487f4bae28/pkg/resource/deploy/step_generator.go#L120

Reply:
Gotcha.